File: libgraffiti/src/css/style.rs
use super::parsing::{fail, ident, list, skip, sym, Parsable, ParseError, Parser};
use super::{BorderStyle, Color, Dimension, Overflow, Px, StyleProp};
use smallbitvec::SmallBitVec;
use std::fmt;
use std::mem::discriminant;

#[derive(Debug, Clone, Default, PartialEq)]
pub struct Style {
    props: Vec<StyleProp>,
    // TODO: inherited_props?
    important_props: SmallBitVec,
}

impl Style {
    pub const EMPTY: Self = Style {
        props: Vec::new(),
        important_props: SmallBitVec::new(),
    };

    pub fn parse(input: &str) -> Result<Self, ParseError> {
        Parsable::parse(input)
    }

    pub fn props(&self) -> &[StyleProp] {
        &self.props
    }

    pub fn apply(&mut self, other: &Style) {
        for (i, p) in other.props.iter().enumerate() {
            self.add_prop(p.clone(), other.important_props[i]);
        }
    }

    fn add_prop(&mut self, prop: StyleProp, important: bool) {
        let d = discriminant(&prop);

        if let Some(i) = self.props.iter().position(|p| discriminant(p) == d) {
            if important || !self.important_props[i] {
                self.props[i] = prop;
                self.important_props.set(i, important);
            }
        } else {
            self.props.push(prop);
            self.important_props.push(important);
        }
    }
}

impl Parsable for Style {
    fn parser<'a>() -> Parser<'a, Self> {
        // any chunk of tokens before ";" or "}"
        let prop_decl = ((!sym(";") * !sym("}") * skip(1)).repeat(1..)).collect() - sym(";").opt();
        let important = || sym("!important").opt().map(|o| o.is_some());

        prop_decl.repeat(0..).map(move |decls| {
            let mut style = Self::default();
            let longhand = StyleProp::parser() + important();
            let shorthand = shorthand_parser() + important();

            for tokens in decls {
                // TODO: shorthands
                if let Ok((prop, important)) = longhand.parse(tokens) {
                    style.add_prop(prop, important);
                }

                if let Ok((props, important)) = shorthand.parse(tokens) {
                    for prop in props {
                        style.add_prop(prop, important);
                    }
                }
            }

            style
        })
    }
}

// impl From<&str> for Style {
//     fn from(s: &str) -> Self {
//         Self::parse(s).unwrap_or_default()
//     }
// }

impl fmt::Display for Style {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        for (i, p) in self.props.iter().enumerate() {
            if i != 0 {
                write!(f, "; ")?;
            }

            write!(f, "{}", p)?;
        }

        Ok(())
    }
}

fn shorthand_parser<'a>() -> Parser<'a, Vec<StyleProp>> {
    use StyleProp::*;

    ident() - sym(":")
        >> |name| match name {
            "background" => background(),
            "flex" => flex(),
            "overflow" => overflow(),
            "outline" => outline(),
            "padding" => rect(PaddingTop, PaddingRight, PaddingBottom, PaddingLeft),
            "margin" => rect(MarginTop, MarginRight, MarginBottom, MarginLeft),
            "border-radius" => rect(
                BorderTopLeftRadius,
                BorderTopRightRadius,
                BorderBottomRightRadius,
                BorderBottomLeftRadius,
            ),
            "border-width" => rect(BorderTopWidth, BorderRightWidth, BorderBottomWidth, BorderLeftWidth),
            "border-style" => rect(BorderTopStyle, BorderRightStyle, BorderBottomStyle, BorderLeftStyle),
            "border-color" => rect(BorderTopColor, BorderRightColor, BorderBottomColor, BorderLeftColor),
            _ => fail("unknown shorthand"),
        }
}

fn rect<'a, T: Parsable + Copy + 'a>(
    top: fn(T) -> StyleProp,
    right: fn(T) -> StyleProp,
    bottom: fn(T) -> StyleProp,
    left: fn(T) -> StyleProp,
) -> Parser<'a, Vec<StyleProp>> {
    list(T::parser(), sym(" ")).convert(move |sides| {
        #[allow(clippy::match_ref_pats)]
        let (a, b, c, d) = match &sides[..] {
            &[a, b, c, d] => (a, b, c, d),
            &[a, b, c] => (a, b, c, b),
            &[a, b] => (a, b, a, b),
            &[a] => (a, a, a, a),
            _ => return Err("expected 1-4 values"),
        };

        Ok(vec![top(a), right(b), bottom(c), left(d)])
    })
}

fn background<'a>() -> Parser<'a, Vec<StyleProp>> {
    (sym("none").map(|_| Color::TRANSPARENT) | Color::parser()).map(|c| vec![StyleProp::BackgroundColor(c)])
}

fn flex<'a>() -> Parser<'a, Vec<StyleProp>> {
    (f32::parser() + (sym(" ") * f32::parser()).opt() + (sym(" ") * Dimension::parser()).opt()).map(
        |((grow, shrink), basis)| {
            vec![
                StyleProp::FlexGrow(grow),
                StyleProp::FlexShrink(shrink.unwrap_or(1.)),
                StyleProp::FlexBasis(basis.unwrap_or(Dimension::ZERO)),
            ]
        },
    )
}

fn overflow<'a>() -> Parser<'a, Vec<StyleProp>> {
    (Overflow::parser() + (sym(" ") * Overflow::parser()).opt())
        .map(|(x, y)| vec![StyleProp::OverflowX(x), StyleProp::OverflowY(y.unwrap_or(x))])
}

fn outline<'a>() -> Parser<'a, Vec<StyleProp>> {
    (Px::parser() + (sym(" ") * BorderStyle::parser()) + (sym(" ") * Color::parser())).map(|((px, style), color)| {
        vec![
            StyleProp::OutlineWidth(px),
            StyleProp::OutlineStyle(style),
            StyleProp::OutlineColor(color),
        ]
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use smallbitvec::sbvec;

    #[test]
    fn parse_style() -> Result<(), ParseError> {
        assert_eq!(Style::parse("")?, Style::EMPTY);
        assert_eq!(Style::parse("unknown-a: 0; unknown-b: 0")?, Style::EMPTY);
        assert_eq!(Style::parse("!important")?, Style::EMPTY);

        assert_eq!(Style::parse("opacity: 0")?.props, &[StyleProp::Opacity(0.)]);

        assert_eq!(
            Style::parse("opacity: 0; opacity: unknown")?.props,
            &[StyleProp::Opacity(0.)]
        );

        assert_eq!(
            Style::parse("opacity: 0; flex-grow: 1")?.props,
            &[StyleProp::Opacity(0.), StyleProp::FlexGrow(1.)]
        );

        assert_eq!(
            Style::parse("opacity: 0 !important")?,
            Style {
                props: vec![StyleProp::Opacity(0.)],
                important_props: sbvec![true]
            }
        );

        Ok(())
    }

    #[test]
    fn parse_shorthands() -> Result<(), ParseError> {
        use super::super::{Color, Dimension, Overflow};
        use StyleProp::*;

        assert_eq!(
            &Style::parse("overflow: hidden")?.props,
            &[OverflowX(Overflow::Hidden), OverflowY(Overflow::Hidden)],
        );

        assert_eq!(
            &Style::parse("overflow: visible hidden")?.props,
            &[OverflowX(Overflow::Visible), OverflowY(Overflow::Hidden)],
        );

        assert_eq!(
            &Style::parse("flex: 1")?.props,
            &[FlexGrow(1.), FlexShrink(1.), FlexBasis(Dimension::ZERO)],
        );

        assert_eq!(
            &Style::parse("flex: 2 3 10px")?.props,
            &[FlexGrow(2.), FlexShrink(3.), FlexBasis(Dimension::Px(10.))],
        );

        assert_eq!(
            &Style::parse("padding: 0")?.props,
            &[
                PaddingTop(Dimension::ZERO),
                PaddingRight(Dimension::ZERO),
                PaddingBottom(Dimension::ZERO),
                PaddingLeft(Dimension::ZERO),
            ],
        );

        assert_eq!(
            &Style::parse("padding: 10px 20px")?.props,
            &[
                PaddingTop(Dimension::Px(10.)),
                PaddingRight(Dimension::Px(20.)),
                PaddingBottom(Dimension::Px(10.)),
                PaddingLeft(Dimension::Px(20.)),
            ],
        );

        assert_eq!(
            &Style::parse("border-radius: 1px 2px 3px 4px")?.props,
            &[
                BorderTopLeftRadius(Px(1.)),
                BorderTopRightRadius(Px(2.)),
                BorderBottomRightRadius(Px(3.)),
                BorderBottomLeftRadius(Px(4.)),
            ],
        );

        assert_eq!(
            &Style::parse("border-radius: 15px 50px")?.props,
            &[
                BorderTopLeftRadius(Px(15.)),
                BorderTopRightRadius(Px(50.)),
                BorderBottomRightRadius(Px(15.)),
                BorderBottomLeftRadius(Px(50.)),
            ],
        );

        assert_eq!(
            &Style::parse("background: none")?.props,
            &[StyleProp::BackgroundColor(Color::TRANSPARENT)],
        );
        assert_eq!(
            &Style::parse("background: #000")?.props,
            &[StyleProp::BackgroundColor(Color::BLACK)],
        );

        Ok(())
    }

    #[test]
    fn css_text() {
        let s = Style::parse("display: block").unwrap();
        assert_eq!(s.to_string(), "display: block");
    }

    #[test]
    fn prop_overriding() {
        let mut s = Style::default();

        s.add_prop(StyleProp::Opacity(0.), false);
        s.add_prop(StyleProp::Opacity(1.), false);

        assert_eq!(s.props, vec![StyleProp::Opacity(1.)]);

        s.add_prop(StyleProp::Opacity(0.), true);
        s.add_prop(StyleProp::Opacity(1.), false);

        assert_eq!(s.props, vec![StyleProp::Opacity(0.)]);
    }

    #[test]
    fn apply() {
        let mut s = Style::default();
        s.apply(&Style::parse("display: block; width: 10px; height: 10px").unwrap());
        s.apply(&Style::parse("display: flex; height: 20px !important").unwrap());
        s.apply(&Style::parse("height: 30px").unwrap());
        assert_eq!(s.to_string(), "display: flex; width: 10px; height: 20px");
    }

    #[test]
    #[ignore]
    fn test_size() {
        use super::super::{BoxShadow, Dimension};
        use crate::util::Atom;
        use std::mem::size_of;

        assert_eq!(size_of::<Box<BoxShadow>>(), size_of::<usize>());
        assert_eq!(size_of::<Atom>(), size_of::<usize>());

        assert_eq!(size_of::<Dimension>(), size_of::<(u32, f32)>());

        // TODO: gets broken when Atom<> or Box<> is added
        assert_eq!(size_of::<StyleProp>(), size_of::<(u8, Dimension)>());
    }
}

File: src/style.zig
const std = @import("std");
const emlay = @import("emlay");
const css = @import("css/mod.zig");

// enums
pub const Visibility = enum { visible, hidden };
pub const Overflow = enum { visible, hidden, scroll, auto };
pub const OutlineStyle = enum { none, solid };
pub const BorderStyle = enum { none, solid };

// TODO: figure out where to put these
pub const Color = @import("css/values/color.zig").Color;

// consts
pub const TRANSPARENT: css.Color = .{ .r = 0, .g = 0, .b = 0, .a = 0 };
pub const CURRENT_COLOR: css.Color = .{ .r = 0, .g = 0, .b = 0, .a = 255 }; // TODO: css.Color.current_color (and convert it in element.applyStyle())

pub const LayerStyle = struct {
    visibility: enum { visible, hidden } = .visible,
    opacity: f32 = 1,

    // transform: []const Transform = &.{},

    // overflow
    overflow_x: Overflow = .visible,
    overflow_y: Overflow = .visible,

    // border-radius
    border_top_left_radius: emlay.Dimension = .{ .px = 0 },
    border_top_right_radius: emlay.Dimension = .{ .px = 0 },
    border_bottom_right_radius: emlay.Dimension = .{ .px = 0 },
    border_bottom_left_radius: emlay.Dimension = .{ .px = 0 },

    // outline
    outline_width: emlay.Dimension = .{ .px = 3 },
    outline_style: BorderStyle = .none,
    outline_color: css.Color = CURRENT_COLOR,

    // background
    background_color: css.Color = TRANSPARENT,
    // TODO: background_image: []const BackgroundImage = &.{},

    // border (widths are part of the LayoutStyle)
    border_top_style: BorderStyle = .none,
    border_top_color: css.Color = CURRENT_COLOR,
    border_right_style: BorderStyle = .none,
    border_right_color: css.Color = CURRENT_COLOR,
    border_bottom_style: BorderStyle = .none,
    border_bottom_color: css.Color = CURRENT_COLOR,
    border_left_style: BorderStyle = .none,
    border_left_color: css.Color = CURRENT_COLOR,
};

// supported CSS props
// - all emlay.Style props (but parse css.Dimension which is converted to emlay.Dimension during applyStyle())
// - all LayerStyle props
// - all TextAttr variants (TODO)
pub const StyleProp = blk: {
    var enum_fields: []const std.builtin.Type.EnumField = &.{};
    var union_fields: []const std.builtin.Type.UnionField = &.{};

    for (.{ emlay.Style, LayerStyle }) |T| {
        for (std.meta.fields(T)) |f| {
            enum_fields = enum_fields ++ .{.{
                .name = f.name,
                .value = enum_fields.len,
            }};

            union_fields = union_fields ++ .{.{
                .name = f.name,
                .type = if (f.type == emlay.Dimension) css.Dimension else f.type,
                .alignment = @alignOf(f.type),
            }};
        }
    }

    break :blk @Type(.{ .Union = .{
        .layout = .Auto,
        .tag_type = @Type(.{ .Enum = .{
            .tag_type = std.math.IntFittingRange(0, enum_fields.len - 1),
            .fields = enum_fields,
            .decls = &.{},
            .is_exhaustive = true,
        } }),
        .fields = union_fields,
        .decls = &.{},
    } });
};

// shorthand props, we parse & print them but we always store expanded longhands
// every shorthand is just a struct of longhands which makes it easy to
// expand them into longhands with a simple inline for loop.
//
// we could concat prefix and field name, but there are some edge-cases like `border-radius`
// where the full name is in mixed order (`border-top-left-radius`)
//
// parsing & printing usually works fine thanks to defaults and optionals
// but it's possible to define `parse()` and `format()` for anything custom
pub const Shorthand = union(enum) {
    flex: struct {
        flex_grow: f32 = 0,
        flex_shrink: f32 = 1,
        flex_basis: css.Dimension = .auto, // .{ .percent = 0 }
    },

    padding: struct {
        padding_top: css.Dimension = .{ .px = 0 },
        padding_right: css.Dimension = .{ .px = 0 },
        padding_bottom: css.Dimension = .{ .px = 0 },
        padding_left: css.Dimension = .{ .px = 0 },

        pub usingnamespace RectMixin(@This());
    },

    margin: struct {
        margin_top: css.Dimension = .{ .px = 0 },
        margin_right: css.Dimension = .{ .px = 0 },
        margin_bottom: css.Dimension = .{ .px = 0 },
        margin_left: css.Dimension = .{ .px = 0 },

        pub usingnamespace RectMixin(@This());
    },

    // TODO: this is simplified but the full syntax is crazy, so
    //       we will likely support just some reasonable subset
    background: struct {
        // TODO: background_image: []const BackgroundImage,
        background_color: css.Color = TRANSPARENT,
    },

    // // TODO: this should expand to 12 longhands but it should only
    // //       accept/print 3 values
    // // border: struct {}

    border_width: struct {
        border_top_width: css.Dimension = .{ .px = 3 },
        border_right_width: css.Dimension = .{ .px = 3 },
        border_bottom_width: css.Dimension = .{ .px = 3 },
        border_left_width: css.Dimension = .{ .px = 3 },
    },

    border_style: struct {
        border_top_style: BorderStyle = .none,
        border_right_style: BorderStyle = .none,
        border_bottom_style: BorderStyle = .none,
        border_left_style: BorderStyle = .none,
    },

    border_color: struct {
        border_top_color: css.Color = TRANSPARENT,
        border_right_color: css.Color = TRANSPARENT,
        border_bottom_color: css.Color = TRANSPARENT,
        border_left_color: css.Color = TRANSPARENT,
    },

    border_top: struct {
        border_top_width: css.Dimension = .{ .px = 3 },
        border_top_style: BorderStyle = .none,
        border_top_color: css.Color = TRANSPARENT,
    },

    border_right: struct {
        border_right_width: css.Dimension = .{ .px = 3 },
        border_right_style: BorderStyle = .none,
        border_right_color: css.Color = TRANSPARENT,
    },

    border_bottom: struct {
        border_bottom_width: css.Dimension = .{ .px = 3 },
        border_bottom_style: BorderStyle = .none,
        border_bottom_color: css.Color = TRANSPARENT,
    },

    border_left: struct {
        border_left_width: css.Dimension = .{ .px = 3 },
        border_left_style: BorderStyle = .none,
        border_left_color: css.Color = TRANSPARENT,
    },

    border_radius: struct {
        border_top_left_radius: css.Dimension = .{ .px = 0 },
        border_top_right_radius: css.Dimension = .{ .px = 0 },
        border_bottom_right_radius: css.Dimension = .{ .px = 0 },
        border_bottom_left_radius: css.Dimension = .{ .px = 0 },
    },

    overflow: struct {
        overflow_x: Overflow = .visible,
        overflow_y: Overflow = .visible,
    },

    outline: struct {
        outline_width: css.Dimension = .{ .px = 3 },
        outline_style: BorderStyle,
        outline_color: css.Color = TRANSPARENT,
    },
};

fn RectMixin(comptime T: type) type {
    const fields = std.meta.fields(T);
    const V = fields[0].type;

    return struct {
        pub fn parseWith(parser: *css.Parser) !T {
            var res: T = undefined;
            var vals: [4]V = undefined;
            vals[0] = try parser.parse(V);
            vals[1] = try parser.parse(?V) orelse vals[0];
            vals[2] = try parser.parse(?V) orelse vals[0];
            vals[3] = try parser.parse(?V) orelse vals[1];

            inline for (fields, vals) |f, v| {
                @field(res, f.name) = v;
            }

            return res;
        }
    };
}

pub const StyleDeclaration = css.StyleDeclaration(StyleProp, Shorthand);

pub const StyleSheet = css.StyleSheet(css.StyleRule(StyleDeclaration));
